e3nn/image/convolution.py

```diff
@@ -3,22 +3,32 @@
 
 import torch
 
-from e3nn.non_linearities import rescaled_act
-from e3nn.kernel import Kernel
+from e3nn.non_linearities.rescaled_act import swish
+from e3nn.kernel_mod import FrozenKernel
 from e3nn.radial import CosineBasisModel
 
 
 class Convolution(torch.nn.Module):
-    def __init__(self, Rs_in, Rs_out, size, **kwargs):
+    def __init__(self, Rs_in, Rs_out, size, steps=(1, 1, 1), **kwargs):
         super().__init__()
 
-        R = partial(CosineBasisModel, max_radius=1.0, number_of_basis=(size + 1) // 2, h=50, L=3, act=rescaled_act.relu)
-        self.kernel = Kernel(Rs_in, Rs_out, R, normalization='component')
-        x = torch.linspace(-1, 1, size)
-        self.r = torch.stack(torch.meshgrid(x, x, x), dim=-1)
+        r = torch.linspace(-1, 1, size)
+        x = r * steps[0] / min(steps)
+        x = x[x.abs() <= 1]
+        y = r * steps[1] / min(steps)
+        y = y[y.abs() <= 1]
+        z = r * steps[2] / min(steps)
+        z = z[z.abs() <= 1]
+        r = torch.stack(torch.meshgrid(x, y, z), dim=-1)  # [x, y, z, R^3]
+
+        R = partial(CosineBasisModel, max_radius=1.0, number_of_basis=(size + 1) // 2, h=50, L=3, act=swish)
+        self.kernel = FrozenKernel(Rs_in, Rs_out, R, r, normalization='component')
         self.kwargs = kwargs
 
     def forward(self, features):
-        k = torch.einsum('xyzij->ijxyz', self.kernel(self.r))
-        k.mul_(2 / k[0, 0].numel() ** 0.5)
+        """
+        :param tensor features: [batch, j, x, y, z]
+        """
+        k = torch.einsum('xyzij->ijxyz', self.kernel())
```

Comment:
I'm worries about the speed of this transpose. Is there a way to avoid it in the forward?

Reply:
We can redefine `Kernel` but I prefer if we use this class and measure the time spent to transpose compare to the convolution. If it's less than 1% I will not touch it